### ISSUE TITLE
23.3 seems to not prune partitions for datetime predicate using floating unix timestamp

### ISSUE THREAD

**Issue by jawm** [ClickHouse/ClickHouse#54420]
**Describe the situation**
While making upgrade to 23.3 LTS, some queries become noticeably slower.
After some investigation, it appears that the query planner no longer filters datetime partition expressions if the predicate is comparing with a unix timestamp in a floating point format. This means it can scan much more than it needs to.

Looking forward to the newest LTS 23.8, the issue is not present, so I think whatever was broken has been fixed, but I have not yet found the source. It would be good for such a fix to be backported to v23.3

**How to reproduce**
On versions 22.3, 23.3, 23.8;

Create table and insert some rows:

```
create table repro (timestamp DateTime) engine = MergeTree partition by toYYYYMMDD(timestamp) order by timestamp;

insert into repro values (now()-1),(now()-2);
```

Look at the query plan.
In old LTS, 22.3:
```
efde337dbbe1 :) EXPLAIN indexes = 1 SELECT count() FROM repro WHERE (timestamp > 0.0)

EXPLAIN indexes = 1
SELECT count()
FROM repro
WHERE timestamp > 0.

Query id: 893b18ec-171b-40dd-8a6c-f7f09647551b

┌─explain─────────────────────────────────────────────────────────────────────────┐
│ Expression ((Projection + Before ORDER BY))                                     │
│   Aggregating                                                                   │
│     Expression (Before GROUP BY)                                                │
│       Filter (WHERE)                                                            │
│         SettingQuotaAndLimits (Set limits and quota after reading from storage) │
│           ReadFromMergeTree                                                     │
│           Indexes:                                                              │
│             MinMax                                                              │
│               Condition: true                                                   │
│               Parts: 1/1                                                        │
│               Granules: 1/1                                                     │
│             Partition                                                           │
│               Keys:                                                             │
│                 toYYYYMMDD(timestamp)                                           │
│               Condition: (toYYYYMMDD(timestamp) in [19700101, +Inf))            │
│               Parts: 1/1                                                        │
│               Granules: 1/1                                                     │
│             PrimaryKey                                                          │
│               Condition: true                                                   │
│               Parts: 1/1                                                        │
│               Granules: 1/1                                                     │
└─────────────────────────────────────────────────────────────────────────────────┘

21 rows in set. Elapsed: 0.001 sec.
```

In 23.3

```
1815b617dc1d :) EXPLAIN indexes = 1 SELECT count() FROM repro WHERE (timestamp > 0.0)

EXPLAIN indexes = 1
SELECT count()
FROM repro
WHERE timestamp > 0.

Query id: b44275f0-63ff-4ac5-b73f-a2e4aa68c0c3

┌─explain─────────────────────────────────────┐
│ Expression ((Projection + Before ORDER BY)) │
│   Aggregating                               │
│     Expression (Before GROUP BY)            │
│       Filter (WHERE)                        │
│         ReadFromMergeTree (_local.repro)    │
│         Indexes:                            │
│           MinMax                            │
│             Condition: true                 │
│             Parts: 1/1                      │
│             Granules: 1/1                   │
│           Partition                         │
│             Condition: true                 │
│             Parts: 1/1                      │
│             Granules: 1/1                   │
│           PrimaryKey                        │
│             Condition: true                 │
│             Parts: 1/1                      │
│             Granules: 1/1                   │
└─────────────────────────────────────────────┘

18 rows in set. Elapsed: 0.001 sec.
```

In 23.8
```
71f4a0f42e10 :) EXPLAIN indexes = 1 SELECT count() FROM repro WHERE (timestamp > 0.0)

EXPLAIN indexes = 1
SELECT count()
FROM repro
WHERE timestamp > 0.

Query id: 25c24d4e-3999-49d7-8cea-d276fe3b3b73

┌─explain────────────────────────────────────────────────────────────┐
│ Expression ((Projection + Before ORDER BY))                        │
│   Aggregating                                                      │
│     Expression (Before GROUP BY)                                   │
│       Filter (WHERE)                                               │
│         ReadFromMergeTree (_local.repro)                           │
│         Indexes:                                                   │
│           MinMax                                                   │
│             Condition: true                                        │
│             Parts: 1/1                                             │
│             Granules: 1/1                                          │
│           Partition                                                │
│             Keys:                                                  │
│               toYYYYMMDD(timestamp)                                │
│             Condition: (toYYYYMMDD(timestamp) in [19700101, +Inf)) │
│             Parts: 1/1                                             │
│             Granules: 1/1                                          │
│           PrimaryKey                                               │
│             Condition: true                                        │
│             Parts: 1/1                                             │
│             Granules: 1/1                                          │
└────────────────────────────────────────────────────────────────────┘

20 rows in set. Elapsed: 0.001 sec.
```

**Expected performance**
22.3 and 23.8 both show correctly pruning based on partition
23.3 does not have this filter in the query plan, which causes poor performance

**Additional context**
Given that this isn't happening on 23.8, I'd assume there has been a fix in between these releases, but it must not have been backported. Any assistance in finding the fix would be good, as I need to run the 23.3 LTS for the time being.

**Comment by SaltTan** [comment]
Broken in 22.4 https://fiddle.clickhouse.com/8b4e6321-31a6-4370-8fb2-20a416d358fc
Fixed in 23.6 https://fiddle.clickhouse.com/9afc4be1-a3fd-413a-a445-861e327aba15

**Comment by bharatnc** [comment]
> Broken in 22.4 https://fiddle.clickhouse.com/8b4e6321-31a6-4370-8fb2-20a416d358fc Fixed in 23.6 https://fiddle.clickhouse.com/9afc4be1-a3fd-413a-a445-861e327aba15

@amosbird @KochetovNicolai I tried looking for the PR that fixed this issue but was unable to. Do you happen to be aware of the fix that made this to work again from v23.6 onwards ? Thank you in advance!

**Comment by amosbird** [comment]
@bharatnc It's likely https://github.com/ClickHouse/ClickHouse/pull/49570  and https://github.com/ClickHouse/ClickHouse/pull/51455

**Comment by bharatnc** [comment]
> @bharatnc It's likely #49570 and #51455

Thank you, that's helpful. I will take a look.

**Comment by bharatnc** [comment]
> > @bharatnc It's likely #49570 and #51455
> 
> Thank you, that's helpful. I will take a look.

Just for the record, the last time I checked, this issue seems to be fixed on v23.8-lts and was only present on v23.3-lts. Perhaps it's since been back ported to v23.3-lts probably (I did not check it).
